java/java/src/main/java/org/whispersystems/libsignal/protocol/DecryptionErrorMessage.java

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (C) 2014-2016 Open Whisper Systems
+ *
+ * Licensed according to the LICENSE file in this repository.
+ */
+package org.whispersystems.libsignal.protocol;
+
+import org.signal.client.internal.Native;
+
+import org.whispersystems.libsignal.InvalidMessageException;
+import org.whispersystems.libsignal.ecc.ECPublicKey;
+import org.whispersystems.libsignal.util.guava.Optional;
+
+public final class DecryptionErrorMessage {
+
+  final long handle;
+
+  @Override
+  protected void finalize() {
+     Native.DecryptionErrorMessage_Destroy(this.handle);
+  }
+
+  public long nativeHandle() {
+    return handle;
+  }
+
+  DecryptionErrorMessage(long handle) {
+    this.handle = handle;
+  }
+
+  public DecryptionErrorMessage(byte[] serialized) throws InvalidMessageException {
+    handle = Native.DecryptionErrorMessage_Deserialize(serialized);
+  }
+
+  public static DecryptionErrorMessage forOriginalMessage(byte[] originalBytes, int messageType, long timestamp) {
+    return new DecryptionErrorMessage(
+      Native.DecryptionErrorMessage_ForOriginalMessage(originalBytes, messageType, timestamp));
+  }
+
+  public byte[] serialize() {
+    return Native.DecryptionErrorMessage_GetSerialized(this.handle);
+  }
+
+  public Optional<ECPublicKey> getRatchetKey() {
+    long keyHandle = Native.DecryptionErrorMessage_GetRatchetKey(this.handle);
+    if (keyHandle == 0) {
+      return Optional.absent();
+    } else {
+      return Optional.of(new ECPublicKey(keyHandle));
+    }
+  }
+
+  public long getTimestamp() {
+    return Native.DecryptionErrorMessage_GetTimestamp(this.handle);
+  }
+
+  /// For testing only
+  public static DecryptionErrorMessage extractFromSerializedContent(byte[] serializedContentBytes) throws InvalidMessageException {
+    return new DecryptionErrorMessage(
+      Native.DecryptionErrorMessage_ExtractFromSerializedContent(serializedContentBytes));
```

Comment:
This does seems suspicious that there's a whole native call just for testing.

Reply:
Cost of not having general support for the Content proto in libsignal-client, because the apps currently all have custom handling for what kind of message is inside. In the future I'd personally like to have more of that handled in libsignal-client but it'd make this PR snowball into even more than it already is.

Reply:
Is it possible to at least move this method to the test class?

So no danger of anyone calling from Java?

Reply:
I feel weird putting Native calls outside of the implementation classes. It really ought to be restricted to package visibility, but unfortunately we have multiple packages and our bridge logic doesn't currently support that.